<launch>

    <!-- Launch file for loading all of the nodes required for running the Raspberry Pi
         and Arduino controls -->

    <arg name="use_imu" default="true" />
    <arg name="use_encoders" default="true" />
    <arg name="port" default="/dev/ttyACM0" />
    <arg name="debug" default="false" />


    <!-- Set up connection with Arduino -->
    <!-- The Arduino should be running the code: arduino_sensors.ino -->
    <node pkg="rosserial_python" name="serial_node" type="serial_node.py" respawn="true">
        <param name="port" value="$(arg port)" />
    </node>


    <!-- Begin IMU Nodes -->
    <group if="$(arg use_imu)">
        <group unless="$(arg debug)">
            <rosparam file="$(find sensors)/config/imu_calibration.yaml" command="load" ns="imu/calibration" />
            <node pkg="sensors" name="imu_orient" type="imu_orient.py" />
            <node pkg="sensors" name="imu_read" type="imu_read.py" respawn ="true" />
        </group>
        <group if="$(arg debug)">
            <rosparam file="$(find sensors)/config/imu_calibration.yaml" command="load" ns="imu/calibration" />
            <node pkg="sensors" name="imu_orient" type="imu_orient.py" output="screen" />
            <node pkg="sensors" name="imu_read" type="imu_read.py" output="screen" respawn="true" />
        </group>
    </group>


    <!-- Begin Encoder Nodes -->
    <group if="$(arg use_encoders)">
        <group unless="$(arg debug)">
            <node pkg="sensors" name="encoder_odom" type="encoder_odom.py" />
            <node pkg="sensors" name="encoder_read" type="encoder_read.py" />
        </group>
        <group if="$(arg debug)">
            <node pkg="sensors" name="encoder_odom" type="encoder_odom.py" output="screen" />
            <node pkg="sensors" name="encoder_read" type="encoder_read.py" output="screen" />
        </group>
    </group>
    
    
    <!-- Begin Throttle Controller Node -->
    <group if="$(arg debug)">
        <node pkg="sensors" name="throttle_controller" type="throttle_control.py" output="screen" />
    </group>
    <group unless="$(arg debug)">
        <node pkg="sensors" name="throttle_controller" type="throttle_control.py" />
    </group>
    
    
    <!-- Begin Steering Controller Node -->
    <group if="$(arg debug)">
        <node pkg="sensors" name="steering_controller" type="steering_control.py" output="screen" />
    </group>
    <group unless="$(arg debug)">
        <node pkg="sensors" name="steering_controller" type="steering_control.py" />
    </group>
    
    <!-- Start Controller GUI -->
    <node pkg="sensors" name="control_gui" type="control_gui.py" />

</launch>
